src/rpc/blockchain.cpp

```diff
@@ -2693,21 +2694,42 @@ UniValue CreateUTXOSnapshot(
 
     afile << metadata;
 
+    std::map<uint256, std::vector<std::pair<uint32_t, Coin>>> mapCoins;
     COutPoint key;
+    uint256 last_hash;
     Coin coin;
     unsigned int iter{0};
+    std::vector<std::pair<uint32_t, Coin>> coins;
 
+    auto write_coins_to_file = [&](AutoFile& afile, const uint256& last_hash, const std::vector<std::pair<uint32_t, Coin>>& coins) {
+        afile << last_hash;
+        WriteCompactSize(afile, coins.size());
+        for (auto [n, coin] : coins) {
+            WriteCompactSize(afile, n);
+            afile << coin;
+        }
+    };
+
+    pcursor->GetKey(key);
+    last_hash = key.hash;
     while (pcursor->Valid()) {
         if (iter % 5000 == 0) node.rpc_interruption_point();
         ++iter;
         if (pcursor->GetKey(key) && pcursor->GetValue(coin)) {
-            afile << key;
-            afile << coin;
+            if (key.hash != last_hash) {
+                write_coins_to_file(afile, last_hash, coins);
+                last_hash = key.hash;
+                coins.clear();
+            }
+            coins.emplace_back(key.n, coin);
```

Comment:
slightly unrelated nit: might be useful to assert that we actually wrote the right number of coins (we just take `maybe_stats->coins_count` and wouldn't abort if the actual number of coins written was different becaus the db changed for some reason). I'm sure https://github.com/bitcoin/bitcoin/pull/15606#discussion_r274479369 is correct, but checking it explicitly seems better.

Reply:
I added a counter and check it at the end